R/update-input.R

```diff
@@ -660,7 +668,46 @@ updateSelectizeInput <- function(session, inputId, label = NULL, choices = NULL,
     data.frame(label = lab, value = choices, stringsAsFactors = FALSE)
   } else {
     # slow path
-    as.data.frame(choices, stringsAsFactors = FALSE)
+    list_names <- names(choices)
+    if (is.null(list_names)) list_names <- ""
+
+    choice_list <- mapply(choices, list_names, FUN = function (choice, name) {
+      group <- ""
+      lab <- name
+      if (lab == "") lab <- as.character(choice)
+
+      if (is.list(choice)) {
+        group <- rep(name, length(choice))
+        choice <- unlist(choice)
+
+        if (is.null(names(choice))) {
+          lab <- as.character(choice)
+        } else {
+          lab <- names(choice)
+          empty_names_indices <- lab == ""
+          lab[empty_names_indices] <- as.character(choice[empty_names_indices])
+        }
+      }
+
+      list(
+        label = lab,
+        value = as.character(choice),
+        group = group
+      )
+    }, SIMPLIFY = FALSE)
+
+
+    extract_vector <- function(x, name) {
+      vecs <- lapply(x, `[[`, name)
```

Comment:
The existing implementation of `extract_vector ` was written to work when each list contained vectors of length >= 1. But if it's known that length == 1, and if the type is known ahead of time, then this could be replaced with `vapply`, and the `do.call` won't be necessary. (I think that the vectors are all char vectors, right?)

Reply:
Actually, never mind what I said. If I'm reading the code correctly now, the lists contained in `choice_list` can have variable length, in which case you need the existing `lapply` version.

Reply:
Yes, each list can contain vectors of length >= 1. For example, if `choices <- list(a = 1, B = list(B = 2, C = 4, 5), c = list(3))` then `choice_list` equals:
```
$a
$a$label
[1] "a"

$a$value
[1] "1"

$a$group
[1] ""


$B
$B$label
[1] "B" "C" "5"

$B$value
[1] "2" "4" "5"

$B$group
[1] "B" "B" "B"


$c
$c$label
[1] "3"

$c$value
[1] "3"

$c$group
[1] "c"
```
However, the vectors are all character vectors.